python/cudf/cudf/core/copy_types.py

```diff
@@ -0,0 +1,140 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
```

Comment:
New code after discussion that introduces "validated" types for columns suitable as gather maps and boolean masks. Used by the new indexing routine.